lib/hierarchical.h

```diff
@@ -465,6 +477,117 @@ class HierarchicalRadiosity {
         }
     }
 
+    /**
+     * Refine all links in all nodes.
+     * @return true if links have been refined.
+     */
+    bool refine_links() {
+        bool done = true;
+        for (auto& p : nodes_) {
+            if(!refine_links(p)) {
```

Comment:
Du kannst einen BFS (Queue statt Stack) machen. In dem beim Besuchen packst du jeden Link auf einen Stack. Wenn du fertig bist, erhält der Stack die Links levelweise von unten.